### PR TITLE
Add Focal Loss

### LINKED ISSUE #319: Focal Loss
Focal loss was used to train [RetinaNet](https://arxiv.org/pdf/1708.02002v2.pdf).

From [Papers with code](https://paperswithcode.com/method/focal-loss):
> A Focal Loss function addresses class imbalance during training in tasks like object detection.

Reference Keras Implementation:
https://keras.io/examples/vision/retinanet/#implementing-smooth-l1-loss-and-focal-loss-as-keras-custom-losses 

What do you think about this component? I think that maybe it could also benefit other models.

### PR DESCRIPTION
**Pull request by quantumalaviya**
# What does this PR do?
Draft PR for Focal Loss

Fixes #319 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [x] Did you write any new necessary tests?

## Who can review?

@LukeWood

### CODE REVIEW

**Review by LukeWood — CHANGES_REQUESTED** [review]

**Comment by LukeWood** [comment]
@quantumalaviya lets add unit tests to this

File: keras_cv/losses/focal.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    Standalone usage:
+    ```python
+        y_true = tf.random.uniform([10], 0, maxval=4)
+        y_pred = tf.random.uniform([10], 0, maxval=4)
+        loss = FocalLoss()
+        loss(y_true, y_pred).numpy()
+    ```
+    Usage with the `compile()` API:
+    ```python
+        model.compile(optimizer='adam', loss=keras_cv.losses.FocalLoss())
```
Comment (LukeWood) [inline keras_cv/losses/focal.py:53]: unindent the python code by one level

File: keras_cv/losses/focal.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        num_classes = tf.cast(tf.shape(y_true)[-1], y_pred.dtype)
+        return y_true * (1.0 - self.label_smoothing) + (
+            self.label_smoothing / num_classes
+        )
+
+    def call(self, y_true, y_pred):
+        y_pred = tf.convert_to_tensor(y_pred)
+        y_true = tf.cast(y_true, y_pred.dtype)
+
+        if self.from_logits:
```
Comment (LukeWood) [inline keras_cv/losses/focal.py:82]: can you replace this with:

```
if self.from_logits:
    y_pred = tf.nn.sigmoid(y_pred)
```

and remove the probs variable entirely?

Comment (LukeWood) [reply keras_cv/losses/focal.py:82]: ah, nevermind I see that line 91 re-uses this.  But perhaps it would be more performant to use the probabilities directly since we already compute them with sigmoid here?

**Review by LukeWood — CHANGES_REQUESTED** [review]
minor comments, looks correct logically but haven't deeply verified it.

File: keras_cv/losses/focal.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            y_pred = tf.nn.sigmoid(y_pred)
+
+        if self.label_smoothing:
+            y_true = self._smooth_labels(y_true, y_pred)
+
+        cross_entropy = K.binary_crossentropy(
+            y_true, y_pred, from_logits=self.from_logits
```
Comment (LukeWood) [inline keras_cv/losses/focal.py:89]: it looks like you're invoking from_logits twice.  You already apply sigmoid in line 83, so this should just be false.

File: keras_cv/losses/focal.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        )
+
+        alpha = tf.where(tf.equal(y_true, 1.0), self._alpha, (1.0 - self._alpha))
+        pt = y_true * y_pred + (1.0 - y_true) * (1.0 - y_pred)
+        loss = alpha * tf.pow(1.0 - pt, self._gamma) * cross_entropy
+
+        return tf.reduce_mean(loss, axis=-1)
```
Comment (LukeWood) [inline keras_cv/losses/focal.py:96]: you can actually drop the reduction, Keras does this for you.   So a loss should return a Tensor of size [batch_size].

This is actually required to support sample weight.

File: keras_cv/losses/focal_test.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    def test_output_shape(self):
+        y_true = tf.cast(
+            tf.random.uniform(shape=[2, 5], minval=0, maxval=2, dtype=tf.int32),
+            tf.float32,
+        )
+        y_pred = tf.random.uniform(shape=[2, 5], minval=0, maxval=1, dtype=tf.float32)
+
+        focal_loss = FocalLoss()
+
+        self.assertAllEqual(focal_loss(y_true, y_pred).shape, [])
```
Comment (LukeWood) [inline keras_cv/losses/focal_test.py:30]: you'll need to update the target shape to be [2]

File: keras_cv/losses/focal.py (hunk 4 above)
Comment (LukeWood) [reply keras_cv/losses/focal.py:96]: nevermind, I see you are already doing this.

File: keras_cv/losses/focal.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    """
+
+    def __init__(
+        self,
+        alpha=0.25,
+        gamma=2,
+        from_logits=False,
+        label_smoothing=0,
+        reduction="auto",
+        name="FocalLoss",
```
Comment (LukeWood) [inline keras_cv/losses/focal.py:64]: add **kwargs and propagate to __init__

Comment (LukeWood) [inline keras_cv/losses/focal.py:64]: drop the name argument and just rely on **kwargs

**Review by LukeWood — CHANGES_REQUESTED** [review]
minor changes, but LGTM overall.

File: keras_cv/losses/focal.py (hunk 3 above)
Comment (quantumalaviya) [reply keras_cv/losses/focal.py:89]: That's true. Thanks for catching this!

File: keras_cv/losses/focal_test.py (hunk 5 above)
Comment (quantumalaviya) [reply keras_cv/losses/focal_test.py:30]: When reduction is auto, it aggregates over the dimensions. It could be a 4D array and the output will still have a shape of [].

Comment (quantumalaviya) [reply keras_cv/losses/focal_test.py:30]: If you prefer keeping reduction="none" by default, I can do that. It may go against the other Keras losses though.

File: keras_cv/losses/focal.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    ```
+    """
+
+    def __init__(
+        self,
+        alpha=0.25,
+        gamma=2,
+        from_logits=False,
+        label_smoothing=0,
+        reduction="auto",
```
Comment (LukeWood) [inline keras_cv/losses/focal.py:63]: reduction is already auto by default in keras.losses.Loss, can drop this and rely on **kwargs here too

File: keras_cv/losses/focal_test.py (hunk 5 above)
Comment (LukeWood) [reply keras_cv/losses/focal_test.py:30]: my bad, sorry this comment is outdated

**Review by LukeWood — CHANGES_REQUESTED** [review]
last comment, then LGTM

File: keras_cv/losses/focal.py (hunk 7 above)
Comment (quantumalaviya) [reply keras_cv/losses/focal.py:63]: I thought about this but decided against it going after Keras losses. 

I think letting **kwargs deal with it is more ideal, however.

File: keras_cv/losses/focal.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+        if self.label_smoothing:
+            y_true = self._smooth_labels(y_true)
+
+        cross_entropy = K.binary_crossentropy(
+            y_true, y_pred, from_logits=self.from_logits
+        )
+
+        alpha = tf.where(tf.equal(y_true, 1.0), self._alpha, (1.0 - self._alpha))
+        pt = y_true * y_pred + (1.0 - y_true) * (1.0 - y_pred)
```
Comment (LukeWood) [inline keras_cv/losses/focal.py:86]: y_pred here is still the logits if `from_logits=True`, whereas if `from_logits=False` we will be using non-logits here.

We should apply sigmoid ourselves and pass from_logits=False to K.binary_crossentropy)

File: keras_cv/losses/focal.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        super().__init__(**kwargs)
+        self._alpha = float(alpha)
+        self._gamma = float(gamma)
+        self.from_logits = from_logits
+        self.label_smoothing = label_smoothing
+
+    def _smooth_labels(self, y_true):
+        return y_true * (1.0 - self.label_smoothing) + 0.5 * self.label_smoothing
+
+    def call(self, y_true, y_pred):
```
Comment (LukeWood) [inline keras_cv/losses/focal.py:74]: I think we need some numerical unit tests.  Can we construct some values and manually compute the expected FocalLoss values?

File: keras_cv/losses/focal.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        alpha=0.25,
+        gamma=2,
+        from_logits=False,
+        label_smoothing=0,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self._alpha = float(alpha)
+        self._gamma = float(gamma)
+        self.from_logits = from_logits
```
Comment (LukeWood) [inline keras_cv/losses/focal.py:68]: please add a unit test comparing the result for `from_logits=True` and `from_logits=False`, where in the from_logits=True case we apply sigmoid to the same data.

So basically...

```
y_true = np.random.uniform((8, 10, ...))
y_logits = np.random.uniform((8, 10, ...), min_value=-1000, max_value=1000)

y_pred = softmax(y_logits)

assertEqual(focal_loss_on_logits(y_true, y_logits), focal_loss(y_true, y_pred))
```

**Review by LukeWood — CHANGES_REQUESTED** [review]
Found some bugs that indicates to me that we need more unit tests, please add the suggested unit tests.  Thanks

File: keras_cv/losses/focal.py (hunk 8 above)
Comment (quantumalaviya) [reply keras_cv/losses/focal.py:86]: Ah, true. My bad again.

I'll just convert after the calculation of `binary_crossentropy`.

File: keras_cv/losses/focal.py (hunk 9 above)
Comment (quantumalaviya) [reply keras_cv/losses/focal.py:74]: This is already done for `from_logits=True`. I'll add one for `from_logits=False`.

**Review by LukeWood — APPROVED** [review]
Thank you Sarvagya, this LGTM

**Comment by LukeWood** [comment]
@quantumalaviya can we discuss the exact differences to model garden here a bit?  I'd like to attempt to introduce this loss to be used in their library, so I'd like to make sure we are away of any numerical differences.  Model garden source: https://pastebin.com/UXcRiPCT
